Make this repo a working Python package

yonatank93 opened the pull request:
Here is the TODO list I had for this process:
- [ ] Convert the geodesic/MBAM into a Python package
  - [X] setup.py
  - [ ] Work on the organization of the package (check with Mark)
- [X] Documentation for each functions (numpydoc style)
  - [X] Rename class/functions according to PEP8
- [X] README
  - [X] Installation - pip and source
  - [X] Example - Refer to example folder
  - [X] Attribution - How to cite the papers
  - [X] Licence - MIT
- [X] A wrapper to compute jacobian and Avv using autograd/JAX.
- [X] Unit test
  - [X] Derivative: Compare analytic to finite difference, jacobian, Avv, jax
  - [X] Geodesic: Use a linear model. Compare the geodesic vector to eigenvector.

yonatank93 committed:
DOC: Convert the documentation to numpydoc style

Add .gitignore

yonatank93 committed:
BUG: Typo on the number of predictions

yonatank93 committed:
DOC: Documentation about attributes, change array_like -> np.ndarray

yonatank93 committed:
Change from using `np.dot` to use matrix multiplication operator

yonatank93 committed:
Refactoring the exponential examples

* Apply Black.
* Apply RST.
* Rename the argument of the callback function to distinguish it from
the `geo` object used in the calculation.
* Define the dimension of the problem from the shape of the time
points and the initial parameter.

yonatank93 committed:
Refactoring on the MMR example

* Apply Black
* Apply RST
* Rename the argument of the callback function to distinguish it from
the `geo` object used in the calculation.
* Define the dimension of the problem from the shape of the time
points and the initial parameter.

yonatank93 committed:
Applay Black

yonatank93 committed:
Initial commit to convert repo into a package

yonatank93 committed:
Move Jacobian and Avv wrapper to finite_difference.py

I think this makes more sense since these derivative estimations use
finite difference method. Also, we are planning to add a wrapper to
JAX/autograd to calculation Jacobian and Avv.

yonatank93 committed:
Add requirements in a separate file

And in setup.py, read the requirements from this file. Also, make
setup.py to read README.md to generate the descriptions.

yonatank93 committed:
Change package name MBAM -> mbam

According to PEP8, package's name should be in lowercase.

yonatank93 committed:
Update TODO list

Commit:
Add wrappers for JAX to calculate jacobian and Avv

And add the corresponding tests

yonatank93 committed:
Skip AD tests if JAX is not installed

yonatank93 committed:
Add random seed

yonatank93 committed:
Update the tolerances to loosen the tests

yonatank93 committed:
Rename model.py -> model_exp.py

yonatank93 committed:
Add MIT License

yonatank93 committed:
Update on README

yonatank93 committed:
Add test for geodesic calculation

The test uses a linear model, since this is the simplest case.

yonatank93 commented:
@mktranstrum Do we need to specify M and N when instantiating `Geodesic`? M is not used anywhere in the calculation, and N can be retrieved from the initial parameters (`self.N = len(x)`).

mktranstrum commented on mbam/auto_diff.py:
```diff
@@ -0,0 +1,61 @@
+    """
+    hess = jax.hessian(f)
+
+    def Avv(x, v):
```
Do not calculate Avv this way.  It will not scale well to large models.  You can calculate Avv without calculating the full Hessian.

mktranstrum commented on mbam/utils.py:
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+
+
+def initial_velocity(x, jac, Avv):
```
I think it would be useful to add:
1. An optional integer argument to specify which direction.  e.g., 0 = sloppiest,  1 = second sloppiest, etc.
2. A boolean argument to specify whether the initial direction should be forward/reverse relative to the acceleration.

yonatank93 committed:
Remove M and N from Geodesic.__init__()

M is not used anywhere in the calculation. N can be retrieved from the
initial parameters.

yonatank93 committed:
ENH: Added options to specify index and forward in initial_velocity

yonatank93 committed:
Improve on how to compute Avv with jax

In this version, Avv is calculated by taking second derivative of F(x
+ v*s) with respect to s at s=0.0.

yonatank93 commented:
@mktranstrum I have addressed your comments. Can you take a look at it?

mktranstrum commented:
Did you see the comment about the initial_velocity function?  It doesn't look like you have made those changes yet.
________________________________
From: Yonatan Kurniawan ***@***.***>
Sent: Wednesday, July 6, 2022 5:35 PM
To: mktranstrum/MBAM ***@***.***>
Cc: mktranstrum ***@***.***>; Mention ***@***.***>
Subject: Re: [mktranstrum/MBAM] WIP: Make this repo a working Python package (PR #1)


@mktranstrum<https://github.com/mktranstrum> I have addressed your comments. Can you take a look at it?

—
Reply to this email directly, view it on GitHub<https://github.com/mktranstrum/MBAM/pull/1#issuecomment-1176856825>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACY77GVCVFGKYOWLIXJXCODVSYJ2VANCNFSM52QV3O5Q>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

yonatank93 commented:
I made those changes. I added 2 optional arguments in initial_velocity: i and forward. See this [commit](https://github.com/mktranstrum/MBAM/pull/1/commits/ccfd75069514f0ab616c5716a32be7a7adf6f0dc).
